cpp/src/arrow/extension/fixed_shape_tensor_test.cc

```diff
@@ -0,0 +1,220 @@
+  auto serialized = fst_type->Serialize();
+  ASSERT_OK_AND_ASSIGN(auto deserialized,
+                       fst_type->Deserialize(fst_type->storage_type(), serialized));
+  ASSERT_TRUE(fst_type->Equals(*deserialized));
+}
+
+void CheckDeserializationRaises(const std::shared_ptr<DataType>& storage_type,
+                                const std::string& serialized,
+                                const std::string& expected_message) {
+  auto fst_type = internal::checked_pointer_cast<FixedShapeTensorType>(
+      fixed_shape_tensor(int64(), {3, 4}));
+  EXPECT_RAISES_WITH_MESSAGE_THAT(Invalid, testing::HasSubstr(expected_message),
+                                  fst_type->Deserialize(storage_type, serialized));
+}
+
+TEST_F(TestExtensionType, MetadataSerializationRoundtrip) {
+  CheckSerializationRoundtrip(ext_type_);
+  CheckSerializationRoundtrip(fixed_shape_tensor(value_type_, {}, {}, {}));
+  CheckSerializationRoundtrip(fixed_shape_tensor(value_type_, {0}, {}, {}));
+  CheckSerializationRoundtrip(fixed_shape_tensor(value_type_, {1}, {0}, {"x"}));
+  CheckSerializationRoundtrip(
+      fixed_shape_tensor(value_type_, {256, 256, 3}, {0, 1, 2}, {"H", "W", "C"}));
+  CheckSerializationRoundtrip(
+      fixed_shape_tensor(value_type_, {256, 256, 3}, {2, 0, 1}, {"C", "H", "W"}));
+
+  auto storage_type = fixed_size_list(int64(), 12);
+  CheckDeserializationRaises(boolean(), R"({"shape":[3,4]})",
+                             "Expected FixedSizeList storage type, got bool");
+  CheckDeserializationRaises(storage_type, R"({"dim_names":["x","y"]})",
+                             "Invalid serialized JSON data");
+  CheckDeserializationRaises(storage_type, R"({"shape":(3,4)})",
+                             "Invalid serialized JSON data");
+  CheckDeserializationRaises(storage_type, R"({"shape":[3,4],"permutation":[1,0,2]})",
+                             "Invalid permutation");
+  CheckDeserializationRaises(storage_type, R"({"shape":[3],"dim_names":["x","y"]})",
+                             "Invalid dim_names");
+}
+
+TEST_F(TestExtensionType, RoudtripBatch) {
+  auto exact_ext_type = internal::checked_pointer_cast<FixedShapeTensorType>(ext_type_);
+
+  std::vector<std::shared_ptr<Buffer>> buffers = {nullptr, Buffer::Wrap(values_)};
+  auto arr_data = std::make_shared<ArrayData>(value_type_, values_.size(), buffers, 0, 0);
+  auto arr = std::make_shared<Int64Array>(arr_data);
+  ASSERT_OK_AND_ASSIGN(auto fsla_arr, FixedSizeListArray::FromArrays(arr, cell_type_));
+  auto data = fsla_arr->data();
+  data->type = ext_type_;
+  auto ext_arr = exact_ext_type->MakeArray(data);
+
+  auto ext_metadata =
+      key_value_metadata({{"ARROW:extension:name", exact_ext_type->extension_name()},
+                          {"ARROW:extension:metadata", serialized_}});
+  auto ext_field = field("f0", exact_ext_type, true, ext_metadata);
+  auto batch = RecordBatch::Make(schema({ext_field}), ext_arr->length(), {ext_arr});
```

Comment:
Something I missed before: you are passing in both an extension array and manually the extension metadata. I think it should only be one of both? 
Creating the batch with just the extension array should add this metadata upon serialization to IPC, and should therefore come back as extension array, without having to manually add the key_value_metadata. But you could keep a test where you write a batch with storage type + manual metadata, and ensure it reads back as extension array.

Reply:
Made this into two tests as you suggested.
Note, test now runs `CompareBatch(*batch, *read_batch2, /*compare_metadata=*/false);` as the type won't roundtrip correctly:
```
Left schema: f0: fixed_size_list<item: int64>[12]
Right schema: f0: extension<arrow.fixed_shape_tensor>
```

Reply:
Does roundtriping extension array currently works? Test below fails:
```cpp
  std::shared_ptr<RecordBatch> read_batch;
  auto ext_field = field(/*name=*/"f0", /*type=*/ext_type_, /*nullable=*/true);
  auto batch = RecordBatch::Make(schema({ext_field}), ext_arr->length(), {ext_arr});
  RoundtripBatch(batch, &read_batch);
  CompareBatch(*batch, *read_batch, /*compare_metadata=*/true);
```
With:
```
Left schema: f0: extension<arrow.fixed_shape_tensor>
Right schema: f0: fixed_size_list<item: int64>[12]
-- metadata --
ARROW:extension:metadata: {"shape":[3,4],"dim_names":["x","y"]}
ARROW:extension:name: arrow.fixed_shape_tensor
```

Reply:
> Note, test now runs `CompareBatch(*batch, *read_batch2, /*compare_metadata=*/false);` as the type won't roundtrip correctly:

Yes, for the case where you don't start with an extension type but manual metadata, that is expected

> Does roundtriping extension array currently works? Test below fails:

Hmm, that should work .. 
I don't directly see anything wrong in the test. It is supposed to roundtrip. You have another test that verifies that the type is actually registered, right?

Reply:
But, CI is not actually failing, so it is working fine?

Reply:
CI is not failing because the current test is not really testing extension array roundtrip but rather the storage one: https://github.com/apache/arrow/blob/0d1e9828ac490ede7eb5e140a4bfc858edf5b69c/cpp/src/arrow/extension/fixed_shape_tensor_test.cc#L208-L216

Reply:
> Hmm, that should work .. I don't directly see anything wrong in the test. It is supposed to roundtrip. You have another test that verifies that the type is actually registered, right?

Yes, this one:
https://github.com/apache/arrow/blob/0d1e9828ac490ede7eb5e140a4bfc858edf5b69c/cpp/src/arrow/extension/fixed_shape_tensor_test.cc#L84-L88

Reply:
Just to be sure, can you register the type inside the test before doing the roundtrip and see if the test then passes, just to be sure that's not the issue?

Reply:
We have existing tests in extension_type_test.cc that do such roundtrip tests, so this should certainly work (and so if not, for some reason it is not recognizing the type, or something is going wrong in recreating it)

Reply:
>  I am not sure if this is desired behavior though:

You should compare with a batch that has the extension type instead (eg the batch created just above, if you name both differently)

Reply:
And compare_metadata can be false then

Reply:
That makes sense. Pushed.